### DISCUSSION TITLE
Add optimization / compression for result (united) PDF document from images (jpeg)

### DISCUSSION THREAD

**Discussion by hotenov** [JakubMelka/PDF4QT#90] [Ideas]
@JakubMelka Hi, thanks for your spare time! ;)

Is it possible to add some optimization / compression for united PDF document? (especially if it contains only images, aka "image-2-pdf" conversion)

I'll try to explain my idea on example:

I have a bunch of images which I want to combine into one PDF document. And **_PDF4QT DocPage Organizer_** tool does this conversion quickly, BUT the size of united PDF document is almost the same as all these images _(see the comparison table below)_.

It's not a big problem if I've read this document and deleted it shortly (the size does not matter). But for home archive / library it's a problem, considering that other tools convert much more efficiently (but they are payed 😢 )

So, I attach the `zip` archive with 14 images _(my school algebra classes 😄 )_ in order you can experiment with them.

[demo for PDF4QT (originals) 24.3 MB unpacked.zip](https://github.com/JakubMelka/PDF4QT/files/12588917/demo.for.PDF4QT.originals.24.3.MB.unpacked.zip)

Unpacked images (i.e. originals) occupy 24.3 MB on my disc with NTFS.

I converted (united / combined / merged) them via `PDF4QT DocPage Organizer` and got a PDF file = 22.4 MB
I also converted them via online tool `jpg2pdf.com` and got a PDF file = 11.0 MB.

### Comparison table

|  | PDF4QT DocPage Organizer | jpg2pdf.com |
|---|:---:|:---:|
| Input size | 24.3 MB | 24.3 MB |
| Result size | 22.4 MB | 11.0 MB |
| Compression | 7.82 % | 54.73 % |
| Download PDF | [demo-sample (converted via PDF4QT) 22.4 MB.pdf](https://github.com/JakubMelka/PDF4QT/files/12589205/demo-sample.converted.via.PDF4QT.22.4.MB.pdf) | [demo-sample (converted via jpg2pdf.com) 11.0 MB.pdf](https://github.com/JakubMelka/PDF4QT/files/12589206/demo-sample.converted.via.jpg2pdf.com.11.0.MB.pdf) |

It's **half** ❗ size of PDF4QT's result file. Not bad, right ?! 😄 

I know about PDF format **nothing** and what magic `jpg2pdf.com` does under the hood. As an assumption, they "compressed" (or limit page size) to _A4_ (297 x 210 mm). But I can't see this difference when I'm zooming in both PDF files.

Thus, if there is no difference, why PDF4QT has no option to enable such "optimization"? I would like to have "small" files that I can read as "big" ones.

That would be great to have such feature in **_PDF4QT DocPage Organizer_** tool and I could forget about `jpg2pdf.com`  (or similar online tools) with their annoying free limitations 😞 .

_P.S. Sorry for long read :)_

**Comment by JakubMelka** [comment]
Greetings and thank you for your lengthy post!

PDF4QT DocPage Organizer strives not to manipulate images in any way, as this may not be desired. This means that it maintains lossless quality. Of course, it is possible to add some level of compression, but in this case, I fear it would have to be lossy. I have examined both of the attached files and in the second one (smaller, produced by jpg2pdf.com), the quality is set at 80%. This means that there has been data loss (though unnoticeable to the naked eye).

However, it is a good idea to include some option for compressing these images, both in the PDF4QT DocPage Organizer and in the main PDF4QT Viewer Profi. Do you have any suggestions about how this could be implemented? What would be useful for you as a user?

**Reply by hotenov** [reply]
Thanks for quick reply!

I think you can add a new field "Image quality" _(or something like that)_ with quality value  on **Assemble Documents** window.

With several predefined values in drop-down list (no compression, 90%, 80%, 75%, etc.) or even custom value :). Where `no compression` will be default value (or 100%, but word value in that case is more clear, in my opinion. Or add explanation in the parenthesis, i.e. `100% (no compression)`).
 
![2023-09-12_21-24-55](https://github.com/JakubMelka/PDF4QT/assets/11818507/0c6cafdb-7bc3-4008-b543-55bbcb5196cb)


You can also add a hint (text) that: more quality - bigger result file (size) and that high level quality may be unnoticeable to the naked eye
I think you will write something clever and clear :) I will look forward to this feature! 👍🏻

**Reply by JakubMelka** [reply]
I have added an image compression feature to the PDF4QT editor.

**Comment by dorianmeric** [comment]
I would also love the option to compress the JPEG images that within the resulting file, in order to help control the size of the resulting PDF doc.
Other tools use two user inputs to do this:
1. the max dpi of images (any image with a higher dpi than this input will get downscaled to the set dpi, and any image with a lower dpi is left untouched). Default: 96dpi
2. the JPEG quality (in %). Default: 60%.
I would suggest a tick box to activate the bitmap compression (maybe called "Compress Images?", and then these two fields can get activated when the tickbox is.

**Reply by JakubMelka** [reply]
PDF4QT Editor can now compress the images - a new feature was added.
